apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,362 @@
+        projectTokenManager.issue(tokensToSell);
+        uint256 purchaseId = projectTokenManager.assignVested(
+            msg.sender,
+            tokensToSell,
+            startDate,
+            startDate.add(vestingCliffPeriod),
+            startDate.add(vestingCompletePeriod),
+            true /* revokable */
+        );
+        totalDaiRaised = totalDaiRaised.add(_daiToSpend);
+
+        // Keep track of dai spent in this purchase for later refunding.
+        purchases[msg.sender][purchaseId] = _daiToSpend;
+
+        emit TokensPurchased(msg.sender, _daiToSpend, tokensToSell, purchaseId);
+    }
+
+    /**
+    * @notice Refunds a purchase made by `_buyer`, with id `_purchaseId`. Each purchase has a purchase id, and needs to be refunded separately.
+    * @param _buyer address The buyer address to refund.
+    * @param _purchaseId uint256 The purchase id to refund.
+    */
+    function refund(address _buyer, uint256 _purchaseId) public {
+        require(currentSaleState() == SaleState.Refunding, ERROR_INVALID_STATE);
+
+        // Recall how much dai to refund for this purchase.
+        uint256 daiToRefund = purchases[_buyer][_purchaseId];
+        require(daiToRefund > 0, ERROR_NOTHING_TO_REFUND);
+        purchases[_buyer][_purchaseId] = 0;
+
+        // (presale) ~~~> dai ~~~> (buyer)
+        require(daiToken.transfer(_buyer, daiToRefund), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (buyer) ~~~> project tokens ~~~> (Token manager)
+        // Note: this assumes that the buyer didn't transfer any of the vested tokens.
+        // The assumption can be made, considering the imposed restriction of fundingPeriod < vestingCliffPeriod < vestingCompletePeriod.
+        (uint256 tokensSold,,,,,) = projectTokenManager.getVesting(_buyer, _purchaseId);
+        projectTokenManager.revokeVesting(_buyer, _purchaseId);
+
+        // (Token manager) ~~~> project tokens ~~~> (burn 💥)
+        projectTokenManager.burn(address(projectTokenManager), tokensSold);
+
+        emit TokensRefunded(_buyer, daiToRefund, tokensSold, _purchaseId);
+    }
+
+    /**
+    * @notice Closes a sale that has reached the funding goal, sending raised funds to the fundraising pool and the beneficiary address, and initializes the Fundraising app by adding the raised funds as collateral tokens.
+    */
+    function close() public {
+        require(currentSaleState() == SaleState.GoalReached, ERROR_INVALID_STATE);
+
+        // (presale) ~~~> dai ~~~> (beneficiary)
+        uint256 daiForBeneficiary = totalDaiRaised.mul(percentFundingForBeneficiary).div(PPM);
+        require(daiToken.transfer(beneficiaryAddress, daiForBeneficiary), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (presale) ~~~> dai ~~~> (pool)
+        uint256 daiForPool = daiToken.balanceOf(address(this));
+        require(daiToken.transfer(fundraisingPool, daiForPool), ERROR_DAI_TRANSFER_REVERTED);
+
+        // Initialize Fundraising app.
+        fundraisingController.addCollateralToken(
```

Comment:
I think this is unnecessary. The way I see it all collateral tokens should be added upfront by the `fundraising-kit` and not by the presale contract. First of all because it will make it more readable but also because it will simplify the permission system.

Reply:
Great! Pls see my comment below.